File: Data Structures & Algorithms/max-area-of-island/submission-7.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        r=len(grid)
        c=len(grid[0])
        maxArea=0
        seen=set()
        def valid(row,col):
            return 0<=row<r and 0<=col<c and grid[row][col]==1
        directions=[(0,1),(1,0),(-1,0),(0,-1)]
        def dfs(row,col):
            nonlocal area
            for dx, dy in directions:
                newRow=row+dx
                newCol=col+dy
                if (newRow,newCol) not in seen and valid(newRow,newCol):
                    seen.add((newRow,newCol))
                    dfs(newRow,newCol)
                    area+=1
        for i in range(r):
            for j in range(c):
                if grid[i][j]==1 and (i,j) not in seen:
                    area=1
                    seen.add((i,j))
                    dfs(i,j)
                    maxArea=max(area,maxArea)
        return maxArea

File: Data Structures & Algorithms/count-number-of-islands/submission-7.java
class Solution {
    boolean[][] visited;
    public boolean valid(char grid[][],int row, int col){
        int r=grid.length;
        int c=grid[0].length;
        return (0<=row)&&(row<r)&&(0<=col)&&(col<c)&&(grid[row][col]=='1');

    }
    public void dfs(char[][] grid, int row, int col){
        int[][] directions={{0,1},{0,-1},{1,0},{-1,0}};
        for (int[] dir: directions){
            int newRow=row+dir[0];
            int newCol=col+dir[1];
            if ((valid(grid,newRow,newCol)) && (visited[newRow][newCol]==false)){
                visited[newRow][newCol]=true;
                dfs(grid,newRow,newCol);
            }
        }

    }
    public int numIslands(char[][] grid) {
        int cnt=0;
        int r=grid.length;
        int c=grid[0].length;
        visited=new boolean[r][c];
        for (int i=0;i<r;i++){
            for (int j=0;j<c;j++){
                int[] next={i,j};
                if ((valid(grid,i,j))&&(visited[i][j]==false)){
                    cnt++;
                    visited[i][j]=true;
                    dfs(grid,i,j);
                }
            }
        }
        return cnt;
    }
}
